Feature: Dynamic document storage paths

mskg opened the pull request:
<!--
Note: All PRs with code changes should be targeted to the `dev` branch, pure documentation changes can target `main`
-->

## Proposed change

What I like about Paperless is the flexibility to provide a file-based document structure, as this allows for easy consumption of the documents with file-based replication. But, the "one size fits all" solution with `PAPERLESS_FILENAME_FORMAT` is not enough. 

The change is simple: 

- Added storage pathes, where each entry is a different PAPERLESS_FILENAME_FORMAT 
- Each document get's a storage path assigned via existing matching mechanisms (all, exact, regular, fuzzy, auto), but is adjustable by the user at any time
- `file_handling.py` has been modified to use the *per document* filename format
- The implementation is complete, also adding bulk-actions, filtering, proposals to the UI and admin interface. I hope I didn't miss anything

<img width="1236" alt="Screenshot 2022-05-10 at 17 23 14" src="https://user-images.githubusercontent.com/46903097/167665367-c91fc3af-0540-465f-a5d9-b0f442d44107.png">

### Usage

I use specific type of storage layouts for different purposes:

- Everything related to warranty claims is a flat directory with file names like `Warranty/2022-04 Correspondent Title.pdf`

- Insurance communication go into `Insurance/Correspondent/...`, whereas related receipts get sorted by year `Insurance/Correspondent/Year/...`

This also allows easy file-based consumption of related documents. The maintenance/assignment gets a no-brainer with auto learning enabled.

### Compatibility 

The change is non-intrusive and fully backward compatible. 

Not using storage pathes or not having one assigned forces the system into using `PAPERLESS_FILENAME_FORMAT`.

### Known limitations

- If an existing storage path get's adjusted, this can only be fixed by running `document_renamer` again
- There is a breaking change implemented regarding file naming. The implementation removes empty pathes and parts of the filename that would otherwise resolve to `none`. This is feature-flagged with PAPERLESS_FILENAME_REMOVE_NONE, defaulting to `No` and such being backward compatible with the existing behaviour.
 
## Type of change

<!--
What type of change does your PR introduce to Paperless-ngx?
NOTE: Please check only one box!
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [X] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [X] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [X] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [X] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [X] I have checked my modifications for any breaking changes.

Commit:
Added devcontainer

Commit:
Add feature storage pathes

Commit:
Exclude tests and add versioning

Commit:
Check escaping

Commit:
Check escaping

Commit:
Check quoting

Commit:
Echo

Commit:
Escape

Commit:
Escape :

Commit:
Double escape \

Commit:
Escaping

Commit:
Remove if

Commit:
Escape colon

Commit:
Missing \

Commit:
Esacpe :

Commit:
Escape all

Commit:
test

Commit:
Remove sed

Commit:
Fix exclude

Commit:
Remove SED command

Commit:
Add LD_LIBRARY_PATH

Commit:
Adjusted to v1.7

Commit:
Adjusted to v1.7

Commit:
Updated test-cases

Commit:
Updated test-cases

Commit:
Remove devcontainer

Commit:
Removed internal build-file

Commit:
Run pre-commit

Commit:
Corrected flak8 error

shamoon commented:
Thanks for the PR! We will take a look asap. In the meantime this should be targeted to `dev` and rebased on that branch.

shamoon commented:
This was closed automatically because it was targeting that outdated branch, I will change it to dev

mskg commented:
I will have a look asap.

Commit:
Corrected flak8 error

mskg commented:
It was a long day ... I rebased it on dev

mskg commented:
> As for the PR itself, I was waiting for others to chime in. Personally, I dont think this is a feature I would use. My preference is to just store docs in some simple way and let paperless handle things from there rather than complicate the file storage. I bet there are some that would, perhaps a vocal group, but IMHO if its not going to get very wide use I dont think I would be in favor of including it as-is since its so prominent, etc (in that case perhaps it could be made optional or something like that?).

It's non intrusive. If you don't use storage pathes the system behaves as before. It's like not using correspondents or document types.

qcasey commented:
Thank you for this PR!

FWIW I would personally find this useful. I agree with this:

> But, the "one size fits all" solution with PAPERLESS_FILENAME_FORMAT is not enough.

Regarding the prominence, the tab might feel more intuitive in the settings panel. Although I haven't tested this to confirm.

As shamoon mentioned this is a change that would introduce more long-term maintenance and testing. I think we should leave this up for a while so others can chime in before deciding to continue working and merge. I can help with documentation at that point :)

qcasey commented:
Thank you for this PR!

FWIW I would personally find this useful. I agree with this:

> But, the "one size fits all" solution with PAPERLESS_FILENAME_FORMAT is not enough.

Regarding the prominence, the tab might feel more intuitive in the settings panel. Although I haven't tested this to confirm.

As shamoon mentioned this is a change that would introduce more long-term maintenance and testing. I think we should leave this up for a while so others can chime in before deciding to continue working and merge. I can help with documentation at that point :)

shamoon commented:
Yea I dont think maintenance is so tough, I do see its largely cloned.

One thing about this at the moment is that the way this is written storage path is treated like document metadata, like something one can search on, filter by, etc. But do you really want that? Is that something people would actually want to do in the way they do for Tags, Correspondents, etc? Perhaps instead they could be tied to metadata, like tied to a correspondent / doc type / tag. Of course thats more complicated and might get messy.

Also the UI would definitely need changes to accommodate another top-level thing like this (that is the filter editor, document cards), Im not sure just adding them on like this will work.

Anyway, yea, lets see.

mskg commented:
> One thing about this at the moment is that the way this is written storage path is treated like document metadata, like something one can search on, filter by, etc. But do you really want that? Is that something people would actually want to do in the way they do for Tags, Correspondents, etc? Perhaps instead they could be tied to metadata, like tied to a correspondent / doc type / tag. Of course thats more complicated and might get messy.

I actually implemented this (attached to correspondents, ...) and threw it away afterwards. It doesn't reflect reality: I put a document somewhere, not a correspondent, or a tag. That's again a simplification that tries to mimic behavior that is not related. 
Afterwards I googled some larger document management systems and looked up how they represent hierarchical structures. They make that relation independent of the other metadata - what makes totally sense. This is why I came up with a per document approach.

And yes, I want to be able to search for it, as it most likely requires bulk-change at some point in time. And I can give you many more cases.

I think it's an elegant solution, not too complicated and reuses all patterns of the existing product. 
For me the change is vital, but most likely I'm biased ;)

Commit:
Merge remote-tracking branch 'origin/dev' into feature-document-storage

Commit:
Merge remote-tracking branch 'origin/dev' into feature-document-storage

Commit:
Adjusted to new plural translations

mskg commented:
How do we want to continue here?

stumpylog commented:
I'm not sure yet if I would use the feature, but I'm alright with including it for others.  The backend changes at least aren't too much.

stumpylog reviewed (changes requested):
I can't give any feedback on the frontend, but I do have a few questions/feedback for the backend.

mskg commented on src/paperless/settings.py:
```diff
@@ -588,6 +588,7 @@ def default_threads_per_worker(task_workers) -> int:
 # TODO: this should not have a prefix.
 # Specify the filename format for out files
 PAPERLESS_FILENAME_FORMAT = os.getenv("PAPERLESS_FILENAME_FORMAT")
+PAPERLESS_FILENAME_REMOVE_NONE = __get_boolean("PAPERLESS_FILENAME_REMOVE_NONE", "NO")
```
I added the documentation. But for the `PAPERLESS_` prefix I just followed the existing code; it's with prefix.

stumpylog replied:
The TODO on line 588 indicates to me it was going to change to remove `PAPERLESS_` and most settings already strip that prefix.

Docs wise, I'd like to see the setting added to `docs\configuration.rst`.  Then the details on dynamic paths in `docs\advanced_usage.rst`, under the "File name handling" heading probably

mskg replied:
> The TODO on line 588 indicates to me it was going to change to remove PAPERLESS_ and most settings already strip that prefix.

I did the change and will create a first draft of the document.

mskg commented on src/documents/tasks.py:
```diff
@@ -52,6 +53,7 @@ def train_classifier():
         not Tag.objects.filter(matching_algorithm=Tag.MATCH_AUTO).exists()
         and not DocumentType.objects.filter(matching_algorithm=Tag.MATCH_AUTO).exists()
         and not Correspondent.objects.filter(matching_algorithm=Tag.MATCH_AUTO).exists()
+        and not StoragePath.objects.exists()
```
Good spot. I initially did the implementation with just the AUTO mechanism and forgot to adjust this part.

mskg commented on src/documents/tests/test_classifier.py:
```diff
@@ -177,6 +192,14 @@ def testSaveClassifier(self):
         new_classifier.load()
         self.assertFalse(new_classifier.train())
 
+    # @override_settings(
```
When you change the file format you need to regenerate the checked-in pickle file. There was no code-path to do that, such I quickly added it as a  test. I left it for those who come. 

I wasn't completely sure if this was done by purpose, so I didn't want to refactor the code as `setup.` 
Comments welcome.

mskg commented on src/documents/file_handling.py:
```diff
@@ -127,13 +127,35 @@ def generate_unique_filename(doc, archive_filename=False):
+        if doc.storage_path:
+            if doc.storage_path.path is not None:
+                logger.debug(
+                    "Document has storage_path %i (%s) set",
```
Agreed. Only did that, because the commit hooks shouted at me ;)

mskg commented on src/documents/file_handling.py:
```diff
@@ -127,13 +127,35 @@ def generate_unique_filename(doc, archive_filename=False):
+                filename_format = doc.storage_path.path
+            else:
+                logger.warn(
+                    "storage_path %i does not have a path set",
```
same as above.

Commit:
Small adjustments due to code-review backend

Commit:
Merged with dev changes from #942

Commit:
Adjusted line-break

tooomm commented on src/paperless/settings.py:
```diff
@@ -589,6 +589,10 @@ def default_threads_per_worker(task_workers) -> int:
 # Specify the filename format for out files
 PAPERLESS_FILENAME_FORMAT = os.getenv("PAPERLESS_FILENAME_FORMAT")
 
+# If this is enabled, variables in filename format with resolve to empty-string instead of none.
```
Typo? "...variables in filename format **will** resolve to..."?

mskg replied:
Updated

fboie commented:
This new feature might solve for quite a few other approaches like e.g. "hierarchical tags".

While I love paperless and being able to find stuff in seconds, reality is that I am not alone. Thus I would like my documents to be filed in at least some meaningful way, so that if get hit by a bus and my docker world dies, my family can still find things.

In the UI world, you design for 50% searchers and 50% browsers. Maybe in paperless the numbers are different, but I do see more than 10 people using this feature. And if the code change and thus maintainance effort is small, I would go for it.

Commit:
Removed PAPERLESS prefix from settings variables

Commit:
Corrected style change due to search+replace

Commit:
First documentation draft

mskg commented:
I added a first draft for the documentation. Comments welcome, I'm not a native speaker.

Also included is a new doc Makefile target `livehtml` that does exactly that: Preview with live reload. I added `sphinx-autobuild` to the dev dependencies for that.

Commit:
Merge remote-tracking branch 'origin/dev' into feature-document-storage

Commit:
Merge remote-tracking branch 'origin/dev' into feature-document-storage

stumpylog commented on Pipfile.lock:
```diff
@@ -739,47 +737,47 @@
             ],
             "index": "pypi",
-            "version": "==9.1.0"
+            "version": "==9.1.1"
```
One thing I really don't like about pipenv is how hard it is to add a single new thing, while not updating everything around it.

Could you redo the addition of the new dep using `pipenv install --keep-outdated sphinx-autobuild`?  That should minimize the changes to this file, including the bump in Pillow version

stumpylog replied:
Thanks, it's an annoyance for sure.

Commit:
Revert changes to Pipfile

stumpylog commented on src/documents/file_handling.py:
```diff
@@ -144,26 +162,26 @@ def generate_filename(doc, counter=0, append_gpg=True, archive_filename=False):
             local_created = timezone.localdate(doc.created)
 
-            path = settings.PAPERLESS_FILENAME_FORMAT.format(
+            path = settings.FILENAME_FORMAT.format(
```
I think this got borked a bit in the merge.  Should be `filename_format` I think

mskg replied:
Don't get it

stumpylog replied:
As this, this will use the global filename format, instead of maybe the document specific format, as grabbed on line 140

mskg replied:
Of course! Will change.

Commit:
Add sphinx-autobuild with keep-outdated

Commit:
Revert merge error that results in wrong storage path is evaluated

Commit:
Adjust styles of generated files ...

stumpylog commented on src/documents/file_handling.py:
```diff
@@ -128,13 +128,31 @@ def generate_unique_filename(doc, archive_filename=False):
-        if settings.PAPERLESS_FILENAME_FORMAT is not None:
-            tags = defaultdictNoStr(lambda: slugify(None), many_to_dictionary(doc.tags))
+        if doc.storage_path:
+            if doc.storage_path.path is not None:
```
Is this condition possible?  I believe the model doesn't allow a null for the field.

stumpylog committed:
Adds additional testing to cover dynamic storage path functionality

Commit:
Merge remote-tracking branch 'origin/dev' into feature-document-storage

Commit:
Remove unnecessary condition

mskg commented:
I'm very happy to see the project being active again. A lot changes to be merged down every day but that's definitely a 👍

stumpylog reviewed (approved):
I manually built the latest into an image and tested it out.  It all looks to work.  As before, can't speak to anything frontend wise, but it does appear ok.

tooomm commented:
> Not sure if I can change it, but we should probably add a requirement of at least one other reviewer.

An admin should be able to adjust the branch protection settings: https://github.blog/2018-03-23-require-multiple-reviewers/

shamoon commented:
Yes this is not ready to merge from a frontend standpoint. I have a bunch of thoughts but I dont want to ask for changes in case this doesnt end up getting merged. Unless youre ok with that @mskg

shamoon commented on src-ui/src/app/app-routing.module.ts:
```diff
@@ -27,6 +28,7 @@ const routes: Routes = [
       { path: 'tags', component: TagListComponent },
       { path: 'documenttypes', component: DocumentTypeListComponent },
       { path: 'correspondents', component: CorrespondentListComponent },
+      { path: 'storagepathes', component: StoragePathListComponent },
```
```suggestion
      { path: 'storagepaths', component: StoragePathListComponent },
```

shamoon commented on src-ui/src/app/components/app-frame/app-frame.component.html:
```diff
@@ -134,6 +134,13 @@ <h6 class="sidebar-heading px-3 mt-4 mb-1 text-muted">
+            <a class="nav-link" routerLink="storagepathes" routerLinkActive="active" (click)="closeMenu()">
+              <svg class="sidebaricon" fill="currentColor">
+                <use xlink:href="assets/bootstrap-icons.svg#folder"/>
+              </svg>&nbsp;<ng-container i18n>Storage pathes</ng-container>
```
```suggestion
              </svg>&nbsp;<ng-container i18n>Storage paths</ng-container>
```

shamoon commented on src-ui/src/app/components/app-frame/app-frame.component.html:
```diff
@@ -134,6 +134,13 @@ <h6 class="sidebar-heading px-3 mt-4 mb-1 text-muted">
             </a>
           </li>
+          <li class="nav-item">
+            <a class="nav-link" routerLink="storagepathes" routerLinkActive="active" (click)="closeMenu()">
```
```suggestion
            <a class="nav-link" routerLink="storagepaths" routerLinkActive="active" (click)="closeMenu()">
```

shamoon commented on src-ui/src/app/components/common/edit-dialog/storage-path-edit-dialog/storage-path-edit-dialog.component.html:
```diff
@@ -0,0 +1,20 @@
+  <div class="modal-body">
+
+    <app-input-text i18n-title title="Name" formControlName="name" [error]="error?.name"></app-input-text>
+    <app-input-text i18n-title title="Path" formControlName="path" [error]="error?.path"></app-input-text>
```
There needs to be some hints here on the frontend. There is no way a user would know what theyre expected to put in here.

mskg commented:
This is already a lot of work and needs daily adjustments. Postponing the merge now would be a waste of the last efforts.

Improving it later is totally fine for me - also helping with that. But it's definitely stable enough to merge it down to dev.

shamoon commented on src-ui/src/app/components/common/edit-dialog/storage-path-edit-dialog/storage-path-edit-dialog.component.ts:
```diff
@@ -0,0 +1,40 @@
+  templateUrl: './storage-path-edit-dialog.component.html',
+  styleUrls: ['./storage-path-edit-dialog.component.scss'],
+})
+export class StoragePathEditDialogComponent extends EditDialogComponent<PaperlessStoragePath> {
```
This needs validation. We cannot let a user insert some crazy thing for storage path and let it break the app. It currently allows me to put anything. This probably should be done in the backend so trying to save an invalid path in some way returns an error from the API.

To be clear, validation wouldn't go here in this class, just mentioning here.

mskg replied:
From a security perspective, no change is required. This would be pure usability and validation. We could solve a first shot same as above.

shamoon replied:
Sorry not sure I understand. I do mean from a usability perspective, when possible the UI tries to prevent changing things to nonsensical values. For example changing a tag color to a non hex value, entering incorrect dates, etc. I would see this as the same. The UI should provide "guardrails".

mskg replied:
Of course :-)

My point was: The backend already takes care of not being able to break the file storage which whatever you configure.

shamoon commented:
> This is already a lot of work and needs daily adjustments. Postponing the merge now would be a waste of the last efforts.
> 
> Improving it later is totally fine for me - also helping with that. But it's definitely stable enough to merge it down to dev.

Sorry no I do not agree. This is not complete from an implementation standpoint and I am not ok with merging as is. I dont just mean that the code will merge via git, I mean I am no OK with the frontend implementation.  I will leave a review now.

shamoon commented on src-ui/src/app/components/document-detail/document-detail.component.html:
```diff
@@ -73,6 +73,8 @@
                             (createNew)="createCorrespondent($event)" [suggestions]="suggestions?.correspondents"></app-input-select>
                         <app-input-select [items]="documentTypes" i18n-title title="Document type" formControlName="document_type" [allowNull]="true"
                             (createNew)="createDocumentType($event)" [suggestions]="suggestions?.document_types"></app-input-select>
+                        <app-input-select [items]="storagePathes" i18n-title title="Storage path" formControlName="storage_path" [allowNull]="true"
```
```suggestion
                        <app-input-select [items]="storagePaths" i18n-title title="Storage path" formControlName="storage_path" [allowNull]="true"
```

shamoon commented on src-ui/src/app/components/document-detail/document-detail.component.html:
```diff
@@ -73,6 +73,8 @@
                         <app-input-select [items]="documentTypes" i18n-title title="Document type" formControlName="document_type" [allowNull]="true"
                             (createNew)="createDocumentType($event)" [suggestions]="suggestions?.document_types"></app-input-select>
+                        <app-input-select [items]="storagePathes" i18n-title title="Storage path" formControlName="storage_path" [allowNull]="true"
+                            (createNew)="createStoragePath($event)" [suggestions]="suggestions?.storage_pathes"></app-input-select>
```
```suggestion
                            (createNew)="createStoragePath($event)" [suggestions]="suggestions?.storage_paths"></app-input-select>
```

shamoon commented on src-ui/src/app/components/document-detail/document-detail.component.ts:
```diff
@@ -163,11 +169,17 @@ export class DocumentDetailComponent
+    this.storagePathService
+      .listAll()
+      .pipe(first())
+      .subscribe((result) => (this.storagePathes = result.results))
```
```suggestion
      .subscribe((result) => (this.storagePaths = result.results))
```

shamoon commented on src-ui/src/app/components/document-detail/document-detail.component.ts:
```diff
@@ -336,6 +349,27 @@ export class DocumentDetailComponent
+        switchMap((newStoragePath) => {
+          return this.storagePathService
+            .listAll()
+            .pipe(map((storagePathes) => ({ newStoragePath, storagePathes })))
```
```suggestion
            .pipe(map((storagePaths) => ({ newStoragePath, storagePaths })))
```

shamoon commented on src-ui/src/app/components/document-detail/document-detail.component.ts:
```diff
@@ -336,6 +349,27 @@ export class DocumentDetailComponent
+        })
+      )
+      .pipe(takeUntil(this.unsubscribeNotifier))
+      .subscribe(({ newStoragePath, documentTypes: storagePathes }) => {
```
```suggestion
      .subscribe(({ newStoragePath, documentTypes: storagePaths }) => {
```

shamoon commented on src-ui/src/app/components/document-detail/document-detail.component.ts:
```diff
@@ -336,6 +349,27 @@ export class DocumentDetailComponent
+      )
+      .pipe(takeUntil(this.unsubscribeNotifier))
+      .subscribe(({ newStoragePath, documentTypes: storagePathes }) => {
+        this.storagePathes = storagePathes.results
```
```suggestion
        this.storagePaths = storagePaths.results
```

shamoon commented on src-ui/src/app/components/document-list/bulk-editor/bulk-editor.component.html:
```diff
@@ -53,6 +53,15 @@
         (apply)="setDocumentTypes($event)">
       </app-filterable-dropdown>
+      <app-filterable-dropdown class="me-2 me-md-3" title="Storage path" icon="folder-fill" i18n-title
+        filterPlaceholder="Filter storage pathes" i18n-filterPlaceholder
```
```suggestion
        filterPlaceholder="Filter storage paths" i18n-filterPlaceholder
```

shamoon commented on src-ui/src/app/components/document-list/bulk-editor/bulk-editor.component.html:
```diff
@@ -53,6 +53,15 @@
       </app-filterable-dropdown>
+      <app-filterable-dropdown class="me-2 me-md-3" title="Storage path" icon="folder-fill" i18n-title
+        filterPlaceholder="Filter storage pathes" i18n-filterPlaceholder
+        [items]="storagePathes"
```
```suggestion
        [items]="storagePaths"
```

shamoon commented on src-ui/src/app/components/document-list/bulk-editor/bulk-editor.component.html:
```diff
@@ -53,6 +53,15 @@
+        [editing]="true"
+        [applyOnClose]="applyOnClose"
+        (open)="openStoragePathDropdown()"
+        [(selectionModel)]="storagePathesSelectionModel"
```
```suggestion
        [(selectionModel)]="storagePathsSelectionModel"
```

shamoon commented on src-ui/src/app/components/document-list/bulk-editor/bulk-editor.component.html:
```diff
@@ -53,6 +53,15 @@
+        [applyOnClose]="applyOnClose"
+        (open)="openStoragePathDropdown()"
+        [(selectionModel)]="storagePathesSelectionModel"
+        (apply)="setStoragePathes($event)">
```
```suggestion
        (apply)="setStoragePaths($event)">
```

shamoon commented on src-ui/src/app/components/document-list/bulk-editor/bulk-editor.component.ts:
```diff
@@ -33,10 +35,12 @@ export class BulkEditorComponent {
   tags: PaperlessTag[]
   correspondents: PaperlessCorrespondent[]
   documentTypes: PaperlessDocumentType[]
+  storagePathes: PaperlessStoragePath[]
```
```suggestion
  storagePaths: PaperlessStoragePath[]
```

shamoon commented on src-ui/src/app/components/document-list/bulk-editor/bulk-editor.component.ts:
```diff
@@ -33,10 +35,12 @@ export class BulkEditorComponent {
   tagSelectionModel = new FilterableDropdownSelectionModel()
   correspondentSelectionModel = new FilterableDropdownSelectionModel()
   documentTypeSelectionModel = new FilterableDropdownSelectionModel()
+  storagePathesSelectionModel = new FilterableDropdownSelectionModel()
```
```suggestion
  storagePathsSelectionModel = new FilterableDropdownSelectionModel()
```

shamoon commented on src-ui/src/app/components/document-list/bulk-editor/bulk-editor.component.ts:
```diff
@@ -68,6 +73,9 @@ export class BulkEditorComponent {
       .subscribe((result) => (this.documentTypes = result.results))
+    this.storagePathService
+      .listAll()
+      .subscribe((result) => (this.storagePathes = result.results))
```
```suggestion
      .subscribe((result) => (this.storagePaths = result.results))
```

shamoon commented on src-ui/src/app/components/document-list/bulk-editor/bulk-editor.component.ts:
```diff
@@ -145,6 +153,17 @@ export class BulkEditorComponent {
+      .getSelectionData(Array.from(this.list.selected))
+      .subscribe((s) => {
+        this.applySelectionData(
+          s.selected_storage_pathes,
```
```suggestion
          s.selected_storage_paths,
```

shamoon commented on src-ui/src/app/components/document-list/bulk-editor/bulk-editor.component.ts:
```diff
@@ -145,6 +153,17 @@ export class BulkEditorComponent {
+      .subscribe((s) => {
+        this.applySelectionData(
+          s.selected_storage_pathes,
+          this.storagePathesSelectionModel
```
```suggestion
          this.storagePathsSelectionModel
```

shamoon commented on src-ui/src/app/components/document-list/bulk-editor/bulk-editor.component.ts:
```diff
@@ -299,6 +318,42 @@ export class BulkEditorComponent {
     }
   }
 
+  setStoragePathes(changedDocumentPathes: ChangedItems) {
```
```suggestion
  setStoragePathes(changedDocumentPaths: ChangedItems) {
```

shamoon commented on src-ui/src/app/components/document-detail/document-detail.component.ts:
```diff
@@ -66,13 +69,15 @@ export class DocumentDetailComponent
 
   correspondents: PaperlessCorrespondent[]
   documentTypes: PaperlessDocumentType[]
+  storagePathes: PaperlessStoragePath[]
```
```suggestion
  storagePaths: PaperlessStoragePath[]
```

shamoon commented on src/paperless/urls.py:
```diff
@@ -34,6 +35,7 @@
 api_router.register(r"logs", LogViewSet, basename="logs")
 api_router.register(r"tags", TagViewSet)
 api_router.register(r"saved_views", SavedViewViewSet)
+api_router.register(r"storage_pathes", StoragePathViewSet)
```
```suggestion
api_router.register(r"storage_paths", StoragePathViewSet)
```

shamoon commented on src/documents/views.py:
```diff
@@ -589,6 +602,10 @@ def post(self, request, format=None):
                 "selected_document_types": [
                     {"id": t.id, "document_count": t.document_count} for t in types
                 ],
+                "selected_storage_pathes": [
```
```suggestion
                "selected_storage_paths": [
```

shamoon commented on src/documents/views.py:
```diff
@@ -589,6 +602,10 @@ def post(self, request, format=None):
                 ],
+                "selected_storage_pathes": [
+                    {"id": t.id, "document_count": t.document_count}
+                    for t in storage_pathes
```
```suggestion
                    for t in storage_paths
```

shamoon commented on src/documents/views.py:
```diff
@@ -335,6 +339,9 @@ def suggestions(self, request, pk=None):
                 "document_types": [
                     dt.id for dt in match_document_types(doc, classifier)
                 ],
+                "storage_pathes": [
```
```suggestion
                "storage_paths": [
```

shamoon commented on src-ui/src/app/components/document-list/filter-editor/filter-editor.component.html:
```diff
@@ -40,6 +40,13 @@
           (open)="onDocumentTypeDropdownOpen()"
           (selectionModelChange)="updateRules()"
           [allowSelectNone]="true"></app-filterable-dropdown>
+        <app-filterable-dropdown class="me-2 flex-fill" title="Storage path" icon="folder-fill" i18n-title
+          filterPlaceholder="Filter storage path" i18n-filterPlaceholder
+          [items]="storagePathes"
+          [(selectionModel)]="storagePathSelectionModel"
+          (open)="onStoragePathDropdownOpen()"
+          (selectionModelChange)="updateRules()"
+          [allowSelectNone]="true"></app-filterable-dropdown>
```
```suggestion
```

Im not sure if this should be included in the filter editor? Not sure how others feel but the filter editor is already quite full and I definitely see the importance in the bulk editor (I want to change the storage path of multiple docs easily) but I dont see as much use for it here.

And if it is retained the HTML shoudl be adjusted to move the created/added buttons to a new line probably wrapping the 4 buttons in their own flex div and the added/created in another

shamoon replied:
To clarify the second part I mean so it wraps better on mobile if this button is retained.

mskg replied:
If we hide it, it would still be available through the advanced search via `path:`. Hiding is an option, but probably not the best one.

What about use an icon for the dates, too? It should fit (at least for now).

mskg replied:
What about
- Calendar with arrow left/right 
- Clock with circular arrow around clock-wise, ccw

shamoon replied:
Not sure where you got those, these are ones we use: https://icons.getbootstrap.com

mskg replied:
Forget my comment. You wrote before / after, but you meant created / added. And that’s not obvious with icons.

shamoon replied:
Oh yea sorry thats what I meant yes.

shamoon replied:
Now that I have write access I can do this.

mskg commented:
Like I said, no problem to remove showstoppers. Waiting for your feedback.

shamoon reviewed (changes requested):
A lot of these are just changing the "pathes" to "paths" (since Im doing this on the web I may have missed a few but you get the idea).

I noted a few more significant (nothing insurmountable just more than the spelling change I mean).

Im ok with merging this as it sounds like people will find it useful, I just mean it needs  a little work frontend-wise first.

shamoon commented:
Ps. Definitely appreciate your work on this, I dont mean to minimize that at all.

shamoon commented on src-ui/src/app/components/document-list/bulk-editor/bulk-editor.component.ts:
```diff
@@ -33,10 +35,12 @@ export class BulkEditorComponent {
+  storagePathesSelectionModel = new FilterableDropdownSelectionModel()
   awaitingDownload: boolean
 
   constructor(
```
Oh, by the way I have a concern about this. My understanding is that editing a storage path doesnt actually move it in the backend, is that correct? If so we need to figure out how to handle this because as-is a user would not know this. Presumably there needs to be a very explicit warning somewhere, or we just actually dont allow changing the path via the frontend but that seems like too much of a limitation.

mskg replied:
Agree. Same limitation as of today when you adjust the storage layout. It’s in the docs.

Probably add this as a sentence to the „what to put here“ hin.

shamoon replied:
I added a note to the dialog as I think its not obvious otherwise. We can have a separate discussion about automatically running `document_renamer`, I agree its a separate question.

mskg replied:
👍. And even the task can be added easily by oneself via the admin interface. It doesn’t require a code change.

mskg commented:
> Ps. Definitely appreciate your work on this, I dont mean to gloss over that at all.

No worries. Let’s just focus on the most important points and do the other things as an optimization. We ship and get feedback - there will be more to do.

shamoon commented:
Another broader question I have is: if no new paths are created, presumably they all use the setting from `PAPERLESS_FILENAME_FORMAT` (or the default) but currently the UI doesnt really make that apparent anywhere, which is sort of confusing. Like perhaps we should indicate that to the user. And btw, the UI would need a way to remove the current path and "reset" it to default.

mskg commented:
> Another broader question I have is: if no new paths are created, presumably they all use the setting from `PAPERLESS_FILENAME_FORMAT` (or the default) but currently the UI doesnt really make that apparent anywhere, which is sort of confusing. Like perhaps we should indicate that to the user. 

„Default“ ?

> And btw, the UI would need a way to remove the current path and "reset" it to default.

It‘s there - same as for all inputs - the little ‚X‘.

shamoon commented:
> Let’s just focus on the most important points and do the other things as an optimization. We ship and get feedback - there will be more to do.

Well I do agree that we'll get feedback but I believe these are not small concerns. As this feature actually alters the files on disk and theres no "undo" I do think we need to refine it as much as we can. Yes, its not going to cause data loss but in general the UI doesnt make many changes to files like this (other than actually deleting files).

Dont worry, Im sure there will be feedback after its released! =)

mskg commented:
Can you probably post a quick snippet / ux draft on where to best put the hints in the edit dialog? I don’t know the styling system out of my mind - would have to look up everything.

shamoon commented:
> Can you probably post a quick snippet / ux draft on where to best put the hints in the edit dialog? I don’t know the styling system out of my mind - would have to look up everything.

If you want to give me access to your repo I can take care of this, it requires a couple changes.

qcasey commented:
> If you want to give me access to your repo I can take care of this, it requires a couple changes.

I'm pretty sure we can push to mskg/paperless-ngx. It's easier with the `gh` utility. I'm making a few doc changes now.

shamoon committed:
Add hint to edit storage path dialog

shamoon committed:
Correct spelling of pathes to paths

mskg commented:
> > If you want to give me access to your repo I can take care of this, it requires a couple changes.
> 
> I'm pretty sure we can push to mskg/paperless-ngx. It's easier with the `gh` utility. I'm making a few doc changes now.

You already did by adding some test-cases.

qcasey committed:
Minor documentation tweaks

qcasey committed:
Minor typo

shamoon committed:
improving wrapping of filter editor buttons with new storage path button

shamoon commented:
@mskg I need to cleanup some commits on this, I think I need to do a force-push to your branch if youre OK with that?

shamoon committed:
Update .gitignore

mskg commented:
I would anyway squash the changes?

qcasey commented:
The docs look good so far, thanks for writing those.

I don't think the `document_renamer` moves existing documents into new storage paths. Am I missing something there or is that unimplemented?

shamoon committed:
Fix select border radius in non input-groups

shamoon commented:
> The docs look good so far, thanks for writing those.
> 
> I don't think the `document_renamer` moves existing documents into new storage paths. Am I missing something there or is that unimplemented?

I think it does? E.g. if you edit a storage path with this feature it doesnt move immediately but then running `document_renamer` does move it.

mskg commented:
> I think it does? E.g. if you edit a storage path with this feature it doesnt move immediately but then running `document_renamer` does move it.

It does. It always compares the calculated path with db values and acts if required.

qcasey commented:
> if you **edit** a storage path with this feature it doesnt move immediately but then running document_renamer does move it.

Right, I see that (emphasis mine). However I don't see the documents move when they are uploaded **before** the Storage Path rule. For example:

1. Upload a document containing the words "test"
2. Create a Storage Path rule matching the word "test"
3. Run the `document_renamer`
4. Notice the document is not renamed

The same document is matched and renamed according to the Storage Path rule if I upload it **after** the rule is created.

mskg commented:
> > if you **edit** a storage path with this feature it doesnt move immediately but then running document_renamer does move it.
> 
> Right, I see that (emphasis mine). However I don't see the documents move when they are uploaded **before** the Storage Path rule. For example:
> 
> 1. Upload a document containing the words "test"
> 2. Create a Storage Path rule matching the word "test"
> 3. Run the `document_renamer`
> 4. Notice the document is not renamed
> 
> The same document is matched and renamed according to the Storage Path rule if I upload it **after** the rule is created.

This is only available for tags - the command is called `document_retagger`. 

I would propose to not implement this, as this would require conflict resolution and priorities of rules. That’s not the case for tags - there can be many. 

That’s probably why it also doesn’t exist for correspondents, and types.

qcasey commented:
Also we should probably add how this feature handles precedence to the docs and UI. 

A Document can have multiple Correspondents/Tags/Document Types but only one Storage Path. This feature appears to pick the oldest rule and apply that, regardless of more specific rules created later.

shamoon committed:
Better storage path edit hint

shamoon committed:
Add note to edit storage path dialog re document_renamer

mskg commented:
> Nevermind, I see your comment above. That's news to me, I was under the misconception we could have multiple.

I saw many comments of people wanting to have many correspondents (sender, receivers). Definitely worth thinking about 😉

qcasey reviewed (approved):
Overall this is much more straightforward than I initially expected. Of course I'll defer to shamoon on the remaining UI stuff but I've tested this a bit and I personally don't have any more concerns.

shamoon committed:
Add note to bulk edit storage path re document_renamer

mskg commented on src-ui/src/app/components/document-list/bulk-editor/bulk-editor.component.ts:
```diff
@@ -340,6 +340,10 @@ export class BulkEditorComponent {
       }
+      modal.componentInstance.message +=
+        '<br/><br/><em>' +
+        $localize`Note: changing the storage path does not apply changes to files until you have run the 'document_renamer' utility.` +
```
That’s not correct. Bulk edit touches every file like you would edit it one by one. The renaming is applied directly.

shamoon replied:
Oh I didnt realize that. Then yes I will remove this note. 

But for me this isnt working at the moment. When I bulk edit the files dont move until I run it.

mskg replied:
Than I have to check this. But I cannot do it today.

shamoon replied:
Oh cool, never mind it was because my Q cluster wasn't running and this uses an async task. Ok great I'll remove that note.

shamoon committed:
Rename FILTER_STORAGE_DIRECTORY to PATH

shamoon committed:
Fix broken filter rule parsing

shamoon committed:
Show default storage if unspecified

shamoon committed:
Remove note re storage path on bulk edit

shamoon committed:
Add basic validation of filename variables

shamoon commented:
Ok great, Im pretty happy with these changes, I do think they were important, appreciate your patience. Also after spending some time with this PR I can see how it will probably be useful to others.

My only outstanding concern is about validating the input. On the one hand users can make their own bad decisions =) on the other hand we might want to help prevent people making unintended mistakes. The link to the docs / example in the hint is helpful but especially for variables I still think we should validate them, for example if someone typed `{cerated_year}`. I added a very crude test for this, basically it tries to format the string the way `generate_filename` with dummy replacements just so that it can catch a `KeyError` and then the front can show a helpful error, see screenshot.

Really, thats it! Let me know if theres a better way to do the validation, ofc.

<img width="491" alt="Screen Shot 2022-05-19 at 1 56 59 PM" src="https://user-images.githubusercontent.com/4887959/169403930-ae36be49-93bd-4f3d-b36a-1b254927ddb6.png">

mskg commented:
> Really, thats it! Let me know if theres a better way to do the validation, ofc.

Syntax wise `{bad variables}` are the only real error that would result in the whole formula not being used. If we validate that - 90% of all errors. 

Possible addition: Show him a formatted example while he types. Than we might also catch the unexpected result.

Really appreciate your work 👍

shamoon commented:
Yes that would be cool or I even thought about some kind of visual drag-and-drop tool to build the path but that seems like another PR.

As long as you Python folks are OK with my `validate_path` implementation, then as they say. LGTM 😄
